scripts/validate-inputs.sh

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+# Reusable CI input validator for GitHub Actions
+# Fails fast on unexpected or unsafe input values.
+
+die() { echo "Input validation error: $*" >&2; exit 2; }
+
+# Validate that inputs contain no control characters
+check_no_ctrl() {
+  local name="$1" val
+  val="${!name-}"
+  if printf '%s' "$val" | LC_ALL=C grep -q '[[:cntrl:]]'; then
+    die "$name contains control characters"
+  fi
+}
+
+sanitize() {
+  local input="$1"
+  local sanitized
+  sanitized=$(printf '%s' "$input" | LC_ALL=C tr -cd '[:alnum:]._ +-' )
+  if [[ "$input" != "$sanitized" ]]; then
+    die "Invalid input detected: $input"
+  fi
+}
+
+validate_environment() {
+  local env="$1"
+  case "$env" in
+    dev|staging|prod) ;;
+    *) die "Invalid environment: $env" ;;
+  esac
+}
+
+validate_version() {
+  local version="$1"
+  if ! [[ "$version" =~ ^[0-9]+\.[0-9]+\.[0-9]+(-[0-9A-Za-z.-]+)?(\+[0-9A-Za-z.-]+)?$ ]]; then
+    die "Invalid version: $version"
+  fi
+}
+
+# --- Known inputs ---
+# TEST_FLAGS: allowed values are empty or "--released"
+check_test_flags() {
+  local val="${TEST_FLAGS-}"
```

Comment:
Wouldn't it be simpler to write `local val="${TEST_FLAGS}"` (omit that dash before the closing brace)? What goes wrong if that dash is omitted?

This question applies everywhere that pattern is used.